File: Final Round/antisocial.cpp
// Copyright (c) 2021 kamyu. All rights reserved.

/*
 * Facebook Hacker Cup 2021 Final Round - Antisocial
 * https://www.facebook.com/codingcompetitions/hacker-cup/2021/final-round/problems/E
 *
 * Time:  O(NlogN)
 * Space: O(N)
 *
 */

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <cassert>
#include <cmath>

using namespace std;

// Template: https://github.com/zigui-ps/VoronoiDiagram/blob/master/VoronoiDiagram.cpp
typedef pair<int, int> pii;
typedef pair<double, double> pdd;

const double EPS = 1e-9;
int dcmp(double x){ return x < -EPS? -1 : x > EPS ? 1 : 0; }

double operator / (pdd a,    pdd b){ return a.first * b.second - a.second * b.first; }
pdd    operator * (double b, pdd a){ return pdd(b * a.first, b * a.second); }
pdd    operator + (pdd a,    pdd b){ return pdd(a.first + b.first, a.second + b.second); }
pdd    operator - (pdd a,    pdd b){ return pdd(a.first - b.first, a.second - b.second); }

double sq(double x){ return x*x; }
double size(pdd p){ return hypot(p.first, p.second); }
double sz2(pdd p){ return sq(p.first) + sq(p.second); }
pdd r90(pdd p){ return pdd(-p.second, p.first); }

pdd line_intersect(pdd a, pdd b, pdd u, pdd v){ return u + (((a-u)/b) / (v/b))*v; }
pdd get_circumcenter(pdd p0, pdd p1, pdd p2){
    return line_intersect(0.5 * (p0+p1), r90(p0-p1), 0.5 * (p1+p2), r90(p1-p2));
}

// https://www.youtube.com/watch?v=h_vvP4ah6Ck
double parabola_intersect(pdd left, pdd right, double sweepline){
    /*
    if(dcmp(left.second - right.second) == 0) return (left.first + right.first) / 2.0; /*/
    auto f2 = [](pdd left, pdd right, double sweepline){
        int sign = left.first < right.first ? 1 : -1;
        pdd m = 0.5 * (left+right);
        pdd v = line_intersect(m, r90(right-left), pdd(0, sweepline), pdd(1, 0));
        pdd w = line_intersect(m, r90(left-v), v, left-v);
        double l1 = size(v-w), l2 = sqrt(sq(sweepline-m.second) - sz2(m-w)), l3 = size(left-v);
        return v.first + (m.first - v.first) * l3 / (l1 + sign * l2);
    };
    if(fabs(left.second - right.second) < fabs(left.first - right.first) * EPS) return f2(left, right, sweepline);// */
    int sign = left.second < right.second ? -1 : 1;
    pdd v = line_intersect(left, right-left, pdd(0, sweepline), pdd(1, 0));
    double d1 = sz2(0.5 * (left+right) - v), d2 = sz2(0.5 * (left-right));
    return v.first + sign * sqrt(max(0.0, d1 - d2));
}

class Beachline{
    public:
        struct node{
            node(){}
            node(pdd point, int idx):point(point), idx(idx), end(0), 
                link{0, 0}, par(0), prv(0), nxt(0) {}
            pdd point; int idx; int end;
            node *link[2], *par, *prv, *nxt;
        };
        node *root;
        double sweepline;

        Beachline() : sweepline(-1e20), root(NULL){ }
        inline int dir(node *x){ return x->par->link[0] != x; }

        //     p        n          p            n
        //    / \      / \        / \          / \
        //   n   d => a   p   or a   n   =>   p   d
        //  / \          / \        / \      / \
        // a   b        b   d      c   d    a   c

        void rotate(node *n){
            node *p = n->par;         int d = dir(n);
            p->link[d] = n->link[!d]; if(n->link[!d]) n->link[!d]->par = p;
            n->par = p->par;          if(p->par) p->par->link[dir(p)] = n;
            n->link[!d] = p;          p->par = n;
        }

        void splay(node *x, node *f = NULL){
            while(x->par != f){
                if(x->par->par == f);
                else if(dir(x) == dir(x->par)) rotate(x->par);
                else rotate(x);
                rotate(x);
            }
            if(f == NULL) root = x;
        }

        void insert(node *n, node *p, int d){
            splay(p); node* c = p->link[d];
            n->link[d] = c; if(c) c->par = n;
            p->link[d] = n; n->par = p;

            node *prv = !d?p->prv:p, *nxt = !d?p:p->nxt;
            n->prv = prv;   if(prv) prv->nxt = n;
            n->nxt = nxt;   if(nxt) nxt->prv = n;
        }

        void erase(node* n){
            node *prv = n->prv, *nxt = n->nxt;
            if(!prv && !nxt){ if(n == root) root = NULL; return; }
            n->prv = NULL;   if(prv) prv->nxt = nxt;
            n->nxt = NULL;   if(nxt) nxt->prv = prv;
            splay(n);
            if(!nxt){
                root->par = NULL; n->link[0] = NULL;
                root = prv;
            }
            else{
                splay(nxt, n);     node* c = n->link[0];
                nxt->link[0] = c;  c->par = nxt;         n->link[0] = NULL;
                n->link[1] = NULL; nxt->par = NULL;
                root = nxt;
            }
        }
        bool get_event(node* cur, double &next_sweep){
            if(!cur->prv || !cur->nxt) return false;
            pdd u = r90(cur->point - cur->prv->point);
            pdd v = r90(cur->nxt->point - cur->point);
            if(dcmp(u/v) != 1) return false;
            pdd p = get_circumcenter(cur->point, cur->prv->point, cur->nxt->point);
            next_sweep = p.second + size(p - cur->point);
            return true;
        }
        node* find_beachline(double x){
            node* cur = root;
            while(cur){
                double left = cur->prv ? parabola_intersect(cur->prv->point, cur->point, sweepline) : -1e30;
                double right = cur->nxt ? parabola_intersect(cur->point, cur->nxt->point, sweepline) : 1e30;
                if(left <= x && x <= right){ splay(cur); return cur; }
                cur = cur->link[x > right];
            }
            return NULL;
        }
}; using BeachNode = Beachline::node;

static BeachNode* arr;
static int sz;
static BeachNode* new_node(pdd point, int idx){
    arr[sz] = BeachNode(point, idx);
    return arr + (sz++);
}

struct event{
    event(double sweep, int idx):type(0), sweep(sweep), idx(idx){}
    event(double sweep, BeachNode* cur):type(1), sweep(sweep), prv(cur->prv->idx), cur(cur), nxt(cur->nxt->idx){}
    int type, idx, prv, nxt;
    BeachNode* cur;
    double sweep;
    bool operator>(const event &l)const{ return sweep > l.sweep; }
};

void VoronoiDiagram(vector<pdd> &input, vector<pdd> &vertex, vector<pii> &edge, vector<pii> &area){
    Beachline beachline = Beachline();
    priority_queue<event, vector<event>, greater<event>> events;

    auto add_edge = [&](int u, int v, int a, int b, BeachNode* c1, BeachNode* c2){
        if(c1) c1->end = edge.size()*2;
        if(c2) c2->end = edge.size()*2 + 1;
        edge.emplace_back(u, v);
        area.emplace_back(a, b);
    };
    auto write_edge = [&](int idx, int v){ idx%2 == 0 ? edge[idx/2].first = v : edge[idx/2].second = v; };
    auto add_event = [&](BeachNode* cur){ double nxt; if(beachline.get_event(cur, nxt)) events.emplace(nxt, cur); };

    int n = input.size(), cnt = 0;
    arr = new BeachNode[n*4]; sz = 0;
    sort(input.begin(), input.end(), [](const pdd &l, const pdd &r){
            return l.second != r.second ? l.second < r.second : l.first < r.first;
            });

    BeachNode* tmp = beachline.root = new_node(input[0], 0), *t2;
    for(int i = 1; i < n; i++){
        if(dcmp(input[i].second - input[0].second) == 0){
            add_edge(-1, -1, i-1, i, 0, tmp);
            beachline.insert(t2 = new_node(input[i], i), tmp, 1);
            tmp = t2;
        }
        else events.emplace(input[i].second, i);
    }
    while(events.size()){
        event q = events.top(); events.pop();
        BeachNode *prv, *cur, *nxt, *site;
        int v = vertex.size(), idx = q.idx;
        beachline.sweepline = q.sweep;
        if(q.type == 0){
            pdd point = input[idx];
            cur = beachline.find_beachline(point.first);
            beachline.insert(site = new_node(point, idx), cur, 0);
            beachline.insert(prv = new_node(cur->point, cur->idx), site, 0);
            add_edge(-1, -1, cur->idx, idx, site, prv);
            add_event(prv); add_event(cur);
        }
        else{
            cur = q.cur, prv = cur->prv, nxt = cur->nxt;
            if(!prv || !nxt || prv->idx != q.prv || nxt->idx != q.nxt) continue;
            vertex.push_back(get_circumcenter(prv->point, nxt->point, cur->point));
            write_edge(prv->end, v); write_edge(cur->end, v);
            add_edge(v, -1, prv->idx, nxt->idx, 0, prv);
            beachline.erase(cur);
            add_event(prv); add_event(nxt);
        }
    }
    delete arr;
}
// end of Template

constexpr double INF = numeric_limits<double>::infinity();

double inner_product(const pdd& a, const pdd& b) {
    return a.first * b.first + a.second * b.second;
}

pdd project_point_segment(const pdd& a, const pdd& b, const pdd& c) {
    double r = inner_product(b - a, b - a);
    if (fabs(r) <= EPS) {
        return a;
    }
    r = min(max(inner_product(c - a, b - a) / r, 0.0), 1.0);
    return a + r * (b - a);
}

double ccw(const pdd&a, const pdd&b, const pdd&c) {
    return (b.first - a.first) * (c.second-a.second) - (b.second - a.second) * (c.first - a.first);
}

bool is_inside_segment_incl(const pdd&t, const pdd&a, const pdd&b) {
    return fabs(ccw(t, a, b)) <= EPS && inner_product(a - t, t - b) >= -EPS;
}

bool is_inside_triangle_incl(const pdd& t, const pdd&a, const pdd&b, const pdd&c) {
    if (is_inside_segment_incl(t, a, b) ||
        is_inside_segment_incl(t, b, c) ||
        is_inside_segment_incl(t, c, a)) {
        return true;
    }
    double d1 = ccw(t, a, b), d2 = ccw(t, b, c), d3 = ccw(t, c, a);
    return (d1 > EPS && d2 > EPS && d3 > EPS) || (d1 < -EPS && d2 < -EPS && d3 < -EPS);
}

void process_voronoi_diagrams(
    double XR, double YR,
    const vector<pdd>& key_points,
    vector<pdd> P,
    vector<int> *key_nodes,
    vector<vector<pair<int, double>>> *adj,
    double *result) {

    vector<pdd> vertex;
    vector<pii> edge, area;
    VoronoiDiagram(P, vertex, edge, area);
    (*adj).resize(size(vertex));
    for (int i = 0; i < size(area); ++i) {
        const auto& [e1, e2] = edge[i];
        if (e1 == -1 || e2 == -1) {
            continue;  // infinite edge
        }
        const auto& v1 = vertex[e1];
        const auto& v2 = vertex[e2];
        if (min(v1.first, v2.first) < -EPS || max(v1.first, v2.first) > XR + EPS ||
            min(v1.second, v2.second) < -EPS || max(v1.second, v2.second) > YR + EPS) {
            continue;  // edge outside rectangle
        }
        const auto& [pi1, pi2] = area[i];
        vector<pdd> center_points = {P[pi1], P[pi2]};
        const auto& d = size(center_points[0] - project_point_segment(v1, v2, center_points[1]));
        (*adj)[e1].emplace_back(e2, d);
        (*adj)[e2].emplace_back(e1, d);
        const auto& mid = 0.5 * (center_points[0] + center_points[1]);
        for (int j : {0, 1}) {
            for (const auto& p : center_points) {
                const auto& kp = key_points[j];
                *result = min(*result, size(kp - p));
                if (is_inside_triangle_incl(kp, p, v1, mid)) {
                    (*key_nodes)[j] = e1;
                } else if (is_inside_triangle_incl(kp, p, mid, v2)) {
                    (*key_nodes)[j] = e2;
                }
            }
        }
    }
}

double mst(
    const vector<int>& key_nodes,
    const vector<vector<pair<int, double>>>& adj,
    double result) {

    vector<bool> lookup(size(adj));
    priority_queue<pair<double, int>> max_heap;
    max_heap.emplace(result, key_nodes[0]);
    while (!empty(max_heap)) {
        const auto [dist, u] = max_heap.top();
        max_heap.pop();
        result = min(result, dist);
        if (u == key_nodes[1]) {
            break;
        }
        if (lookup[u]) {
            continue;
        }
        lookup[u] = true;
        for (const auto& [v, dist] : adj[u]) {
            if (lookup[v]) {
                continue;
            }
            max_heap.emplace(dist, v);
        }
    }
    return result;
}

double antisocial() {
    int XR, YR;
    cin >> XR >> YR;
    vector<pdd> key_points(2);
    cin >> key_points[0].first >> key_points[0].second
        >> key_points[1].first >> key_points[1].second;
    int N;
    cin >> N;
    vector<pdd> P;
    for (int i = 0; i < N; ++i) {
        int X, Y;
        cin >> X >> Y;
        P.emplace_back(X, Y);
        P.emplace_back(-X, Y);
        P.emplace_back(X, -Y);
        P.emplace_back(2 * XR - X, Y);
        P.emplace_back(X, 2 * YR - Y);
    }
    vector<int> key_nodes(2, -1);
    vector<vector<pair<int, double>>> adj;
    double result = INF;
    process_voronoi_diagrams(XR, YR, key_points, P, &key_nodes, &adj, &result);
    assert(key_nodes[0] >= 0 && key_nodes[1] >= 0);
    return mst(key_nodes, adj, result);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.precision(10);
    int T;
    cin >> T;
    for (int test = 1; test <= T; ++test) {
        cout << "Case #" << test << ": " << antisocial() << endl;
    }
    return 0;
}

File: Final Round/antisocial.py
# Copyright (c) 2021 kamyu. All rights reserved.
#
# Facebook Hacker Cup 2021 Final Round - Problem E. Antisocial
# https://www.facebook.com/codingcompetitions/hacker-cup/2021/final-round/problems/E
#
# Time:  O(NlogN), TLE in both PyPy2 and Python2
# Space: O(N)
#

from heapq import heappush, heappop

# Template translated from:
# https://github.com/zigui-ps/VoronoiDiagram/blob/master/VoronoiDiagram.cpp
def add(a, b):
    return (a[0] + b[0], a[1] + b[1])

def sub(a, b):
    return (a[0] - b[0], a[1] - b[1])

def mul(b, a):
    return (b * a[0], b * a[1])

def inner_product(a, b):
    return a[0] * b[0] + a[1] * b[1]

def outer_product(a, b):
    return a[0] * b[1] - a[1] * b[0]

def size(p):
    return (p[0]*p[0] + p[1]*p[1])**0.5

def sz2(p):
    return p[0]*p[0] + p[1]*p[1]

def r90(p):
    return (-p[1], p[0])

def dcmp(x):
    return -1 if x < -EPS else 1 if x > EPS else 0

def line_intersect(a, b, u, v):
    return add(u, mul((outer_product(sub(a, u), b) / outer_product(v, b)), v))

def get_circumcenter(p0, p1, p2):
    return line_intersect(mul(0.5, add(p0, p1)), r90(sub(p0, p1)), mul(0.5, add(p1, p2)), r90(sub(p1, p2)))

# https://www.youtube.com/watch?v=h_vvP4ah6Ck
def parabola_intersect(left, right, sweepline):
    if dcmp(left[1] - right[1]) == 0:
        return (left[0] + right[0]) / 2.0

    sign = -1.0 if left[1] < right[1] else 1.0
    v = line_intersect(left, sub(right, left), (0.0, sweepline), (1.0, 0.0))
    d1 = sz2(sub(mul(0.5, add(left, right)), v))
    d2 = sz2(mul(0.5, sub(left, right)))
    return v[0] + sign * ((max(0.0, d1 - d2))**0.5)

class Node(object):
    def __init__(self, point, idx):
        self.point = point
        self.idx = idx
        self.end = 0
        self.link = [None]*2
        self.par = self.prv = self.nxt = None

class Beachline(object):
    def __init__(self):
        self.sweepline = -INF
        self.root = None

    def direction(self, x):
        return x.par.link[0] != x

    def rotate(self, n):
        p = n.par
        d = self.direction(n)
        p.link[d] = n.link[not d]
        if n.link[not d]:
            n.link[not d].par = p
        n.par = p.par
        if p.par:
            p.par.link[self.direction(p)] = n
        n.link[not d] = p
        p.par = n

    def splay(self, x, f=None):
        while x.par != f:
            if x.par.par == f:
                pass
            elif self.direction(x) == self.direction(x.par):
                self.rotate(x.par)
            else:
                self.rotate(x)
            self.rotate(x)
        if f is None:
            self.root = x

    def insert(self, n, p, d):
        self.splay(p)
        c = p.link[d]
        n.link[d] = c
        if c:
            c.par = n
        p.link[d] = n
        n.par = p
        prv = p.prv if not d else p
        nxt = p if not d else p.nxt
        n.prv = prv
        if prv:
            prv.nxt = n
        n.nxt = nxt
        if nxt:
            nxt.prv = n

    def erase(self, n):
        prv, nxt = n.prv, n.nxt
        if (not prv) and (not nxt):
            if n == self.root:
                self.root = None
            return
        n.prv = None
        if prv:
            prv.nxt = nxt
        n.nxt = None
        if nxt:
            nxt.prv = prv
        self.splay(n)
        if not nxt:
            self.root.par = None
            n.link[0] = None
            self.root = prv
        else:
            self.splay(nxt, n)
            c = n.link[0]
            nxt.link[0] = c
            c.par = nxt
            n.link[0] = None
            n.link[1] = None
            nxt.par = None
            self.root = nxt

    def get_event(self, cur, next_sweep):
        if (not cur.prv) or (not cur.nxt):
            return False
        u = r90(sub(cur.point, cur.prv.point))
        v = r90(sub(cur.nxt.point, cur.point))
        if dcmp(outer_product(u, v)) != 1:
            return False
        p = get_circumcenter(cur.point, cur.prv.point, cur.nxt.point)
        next_sweep[0] = p[1] + size(sub(p, cur.point))
        return True

    def find_beachline(self, x):
        cur = self.root
        while cur:
            left = parabola_intersect(cur.prv.point, cur.point, self.sweepline) if cur.prv else -INF
            right = parabola_intersect(cur.point, cur.nxt.point, self.sweepline) if cur.nxt else INF
            if left <= x <= right:
                self.splay(cur)
                return cur
            cur = cur.link[x > right]
        return None

def new_node(arr, point, idx):
    arr.append(Node(point, idx))
    return arr[-1]

def VoronoiDiagram(points):
    vertex, edge, area = [[] for _ in xrange(3)]
    bl = Beachline()
    events = []
    def add_edge(u, v, a, b, c1, c2):
        if c1:
            c1.end = len(edge)*2
        if c2:
            c2.end = len(edge)*2 + 1
        edge.append([u, v])
        area.append((a, b))

    def write_edge(idx, v):
        edge[idx//2][idx%2] = v

    def add_event(cur):
        nxt = [0.0]
        if bl.get_event(cur, nxt):
            heappush(events, (nxt[0], 1, -1, cur.prv.idx, cur.nxt.idx, cur))

    n = len(points)
    arr = []
    points.sort(key=lambda x: (x[1], x[0]))
    tmp = bl.root = new_node(arr, points[0], 0)
    for i in xrange(1, n):
        if dcmp(points[i][1] - points[0][1]) == 0:
            add_edge(-1, -1, i-1, i, 0, tmp)
            t2 = new_node(arr, points[i], i)
            bl.insert(t2, tmp, 1)
            tmp = t2
        else:
            heappush(events, (points[i][1], 0, i, -1, -1, None))
    while events:
        q_sweep, q_type, q_idx, q_prv, q_nxt, q_cur = heappop(events)
        v, idx = len(vertex), q_idx
        bl.sweepline = q_sweep
        if q_type == 0:
            point = points[idx]
            cur = bl.find_beachline(point[0])
            site = new_node(arr, point, idx)
            bl.insert(site, cur, 0)
            prv = new_node(arr, cur.point, cur.idx)
            bl.insert(prv, site, 0)
            add_edge(-1, -1, cur.idx, idx, site, prv)
            add_event(prv)
            add_event(cur)
        else:
            cur, prv, nxt = q_cur, q_cur.prv, q_cur.nxt
            if (not prv) or (not nxt) or (prv.idx != q_prv) or (nxt.idx != q_nxt):
                continue
            vertex.append(get_circumcenter(prv.point, nxt.point, cur.point))
            write_edge(prv.end, v)
            write_edge(cur.end, v)
            add_edge(v, -1, prv.idx, nxt.idx, 0, prv)
            bl.erase(cur)
            add_event(prv)
            add_event(nxt)
    return vertex, edge, area

def project_point_segment(a, b, c):
    r = inner_product(sub(b, a), sub(b, a))
    if abs(r) <= EPS:
        return a
    r = min(max(inner_product(sub(c, a), sub(b, a))/r, 0.0), 1.0)
    return add(a, mul(r, sub(b, a)))

def ccw(a, b, c):
    return (b[0]-a[0])*(c[1]-a[1]) - (b[1]-a[1])*(c[0]-a[0])

def is_inside_segment_incl(t, a, b):
    return abs(ccw(t, a, b)) <= EPS and inner_product(sub(a, t), sub(t, b)) >= -EPS

def is_inside_triangle_incl(t, a, b, c):
    if is_inside_segment_incl(t, a, b) or is_inside_segment_incl(t, b, c) or is_inside_segment_incl(t, c, a):
        return True
    d1, d2, d3 = ccw(t, a, b), ccw(t, b, c), ccw(t, c, a)
    return (d1 > EPS and d2 > EPS and d3 > EPS) or (d1 < -EPS and d2 < -EPS and d3 < -EPS)

def process_voronoi_diagrams(XR, YR, key_points, P):
    vertex, edge, area = VoronoiDiagram(P)
    key_nodes = [-1]*2
    adj = [[] for _ in xrange(len(vertex))]
    result = INF
    for i in xrange(len(area)):
        e1, e2 = edge[i]
        if e1 == -1 or e2 == -1:
            continue  # infinite edge
        v1, v2 = vertex[e1], vertex[e2]
        if (min(v1[0], v2[0]) < -EPS or max(v1[0], v2[0]) > XR + EPS or
            min(v1[1], v2[1]) < -EPS or max(v1[1], v2[1]) > YR + EPS):
            continue  # edge outside rectangle
        pi1, pi2 = area[i]
        sites = [P[pi1], P[pi2]]
        d = size(sub(sites[0], project_point_segment(v1, v2, sites[0])))
        adj[e1].append((e2, d))
        adj[e2].append((e1, d))
        mid = mul(0.5, add(sites[0], sites[1]))
        for j, kp in enumerate(key_points):
            for p in sites:
                result = min(result, size(sub(kp, p)))
                # if kp is inside in both right triangles:
                # - kp is on (p, mid) perpendicular segment
                # - kp and mid are both outside of (p, v1, v2) triangle
                # => it is safe to choose either of v1, v2 since min(|p-v1|, |p-v2|) >= d
                # if kp is inside only one right triangle:
                # => it is safe to choose corresponding v since |p-v| >= d
                if is_inside_triangle_incl(kp, p, v1, mid):
                    key_nodes[j] = e1
                elif is_inside_triangle_incl(kp, p, mid, v2):
                    key_nodes[j] = e2
    return key_nodes, adj, result

def mst(key_nodes, adj, result):
    lookup = [False]*len(adj)
    max_heap = [(-result, key_nodes[0])]
    while max_heap:
        dist, u = heappop(max_heap)
        result = min(result, -dist)
        if u == key_nodes[1]:
            break
        if lookup[u]:
            continue
        lookup[u] = True
        for v, dist in adj[u]:
            if lookup[v]:
                continue
            heappush(max_heap, (-dist, v))
    return result

def antisocial():
    XR, YR = map(float, raw_input().strip().split())
    XA, YA, XB, YB = map(float, raw_input().strip().split())
    key_points = [(XA, YA), (XB, YB)]
    P = []
    for _ in xrange(input()):
        X, Y = map(float, raw_input().strip().split())
        P.append((X, Y))
        P.append((-X, Y))
        P.append((X, -Y))
        P.append((2*XR - X, Y))
        P.append((X, 2*YR - Y))
    key_nodes, adj, result = process_voronoi_diagrams(XR, YR, key_points, P)
    assert(key_nodes[0] >= 0 and key_nodes[1] >= 0)
    return mst(key_nodes, adj, result)

INF = float("inf")
EPS = 1e-9
for case in xrange(input()):
    print 'Case #%d: %s' % (case+1, antisocial())
